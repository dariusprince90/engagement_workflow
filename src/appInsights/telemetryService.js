/* istanbul ignore file -- justification: this file is for configuring application insights */

import { ApplicationInsights } from '@microsoft/applicationinsights-web';
import { ReactPlugin } from '@microsoft/applicationinsights-react-js';

let reactPlugin = null;
let appInsights = null;

const createTelemetryService = () => {
  const initialize = (connectionString) => {
    const errorPrefix = 'Could not initialize app insights telemetry service:';

    if (!connectionString) {
      throw new Error(`${errorPrefix} connection string not provided`);
    }

    reactPlugin = new ReactPlugin();

    appInsights = new ApplicationInsights({
      config: {
        connectionString: connectionString,
        maxBatchInterval: 500,
        disableFetchTracking: false,
        autoTrackPageVisitTime: true,
        extensions: [reactPlugin],
        enableAutoRouteTracking: true
      }
    });

    appInsights.loadAppInsights();
  };

  return { reactPlugin, appInsights, initialize };
};

export const telemetryService = createTelemetryService();
export const getAppInsights = () => appInsights;
