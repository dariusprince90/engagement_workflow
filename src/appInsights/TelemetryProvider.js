import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { withAITracking } from '@microsoft/applicationinsights-react-js';

import * as config from '../configs/config';
import { telemetryService } from './telemetryService';

const propTypes = {
  children: PropTypes.element.isRequired
};

export const TelemetryProvider = ({ children }) => {
  // **********************************************************************
  // * constants

  const [isInitialized, setIsInitialized] = useState(false);

  // **********************************************************************
  // * component vars

  // **********************************************************************
  // * functions

  // **********************************************************************
  // * side effects

  useEffect(() => {
    const appInsightsConnectionString = config.AppInsights.connectionString;

    if (!Boolean(isInitialized) && Boolean(appInsightsConnectionString)) {
      telemetryService.initialize(appInsightsConnectionString);
      setIsInitialized(true);
    }
  }, [isInitialized]);

  // **********************************************************************
  // * render

  return <>{children}</>;
};

TelemetryProvider.propTypes = propTypes;

export default withAITracking(telemetryService.reactPlugin, TelemetryProvider);
