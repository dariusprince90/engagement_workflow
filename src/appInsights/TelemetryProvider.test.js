import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import * as reactRouterDom from 'react-router-dom';

import { TelemetryProvider } from './TelemetryProvider';
import { telemetryService } from './telemetryService';

// **********************************************************************
// * constants

const FakeChildren = () => <fake-children />;

const getComponentToRender = () => (
  <TelemetryProvider>
    <FakeChildren />
  </TelemetryProvider>
);

// **********************************************************************
// * mock external dependencies

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn()
}));

jest.mock('./telemetryService', () => {
  return {
    telemetryService: {
      initialize: jest.fn()
    }
  };
});

// **********************************************************************
// * unit tests

describe('TelemetryProvider', () => {
  // **********************************************************************
  // * setup

  beforeAll(() => {
    jest.spyOn(telemetryService, 'initialize');
    jest.spyOn(reactRouterDom, 'useHistory');
  });

  // **********************************************************************
  // * tear-down

  afterEach(() => {
    telemetryService.initialize.mockClear();
    reactRouterDom.useHistory.mockClear();
  });

  // **********************************************************************
  // * execution

  describe('rendering', () => {
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(getComponentToRender(), div);
    });

    it('renders the children', () => {
      const tree = renderer.create(getComponentToRender()).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('functional', () => {
    it('initializes the telemetry service when history is defined', () => {
      reactRouterDom.useHistory.mockReturnValue();
      render(getComponentToRender());
      expect(telemetryService.initialize).toHaveBeenCalled();
    });
  });
});
