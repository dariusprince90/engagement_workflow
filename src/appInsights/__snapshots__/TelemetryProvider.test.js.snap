// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TelemetryProvider rendering renders the children 1`] = `<fake-children />`;
